repeater/src/main/scala/com.snowplowanalytics.snowplow.storage.bigquery.repeater/Resources.scala

```diff
@@ -26,40 +27,43 @@ import io.chrisdavenport.log4cats.Logger
 import fs2.{Chunk, Stream}
 import fs2.concurrent.{Queue, SignallingRef}
 
-import com.snowplowanalytics.snowplow.storage.bigquery.common.Config
 import com.snowplowanalytics.snowplow.badrows.BadRow
-import com.snowplowanalytics.snowplow.storage.bigquery.repeater.RepeaterCli.GcsPath
+import com.snowplowanalytics.snowplow.storage.bigquery.common.config.CliConfig.Environment.RepeaterEnvironment
+import com.snowplowanalytics.snowplow.storage.bigquery.repeater.RepeaterCli.{GcsPath, validateBucket}
 
 /**
-  * Resources container, allowing to manipulate all acquired entities
   * @param bigQuery BigQuery client provided by SDK
-  * @param bucket destination path
-  * @param env whole configuration, extracted from config files
-  * @param desperates queue of events that still could not be loaded into BQ
-  * @param counter counter for batches of desperates
-  * @param stop a signal to stop retreving items
-  * @param concurrency a number of streams to execute inserts
+  * @param bucket Destination for uninsertable events
+  * @param env App configuration extracted from config files
+  * @param uninsertable Queue of events that could not be loaded into BQ after multiple attempts
+  * @param counter Counter for batches of uninsertable events
+  * @param stop Signal to stop retrieving items
+  * @param statistics Holds information about inserted and uninsertable events over time
+  * @param bufferSize Max size of uninsertable buffer (buffer is flushed when this is reached)
+  * @param timeout Max time to wait before flushing the uninsetable buffer (if bufferSize is not reached first)
+  * @param backoffPeriod Time to wait before trying to re-insert events
+  * @param concurrency Number of streams to execute inserts
   */
-class Resources[F[_]: Sync](
+final class Resources[F[_]: Sync](
   val bigQuery: BigQuery,
   val bucket: GcsPath,
-  val env: Config.Environment,
-  val desperates: Queue[F, BadRow],
+  val env: RepeaterEnvironment,
+  val uninsertable: Queue[F, BadRow],
   val counter: Ref[F, Int],
   val stop: SignallingRef[F, Boolean],
   val statistics: Ref[F, Resources.Statistics],
   val bufferSize: Int,
-  val windowTime: Int,
-  val backoffTime: Int,
+  val timeout: Int,
+  val backoffPeriod: Int,
   val concurrency: Int,
   val insertBlocker: Blocker,
   val jobStartTime: Instant
 ) {
-  def logInserted: F[Unit] =
+  def logInserted(): F[Unit] =
```

Comment:
I'm not going to argue about this one.... but why the opinionated change from no parentheses to parentheses?

Reply:
That was an IDE suggestion, I just assumed it was about some style convention to clearly flag `Unit` return types.

Reply:
Apparently, it's in the Programming in Scala book:

> In principle it's possible to leave out all empty parentheses in Scala function calls. However, it is recommended to still write the empty parentheses when the invoked method represents more than a property of its receiver object. For instance, empty parentheses are appropriate if the method performs I/O, or writes reassignable variables (vars), or reads vars other than the receiver's fields, either directly or indirectly by using mutable objects. That way, the parameter list acts as a visual clue that some interesting computation is triggered by the call. For instance:

```
  "hello".length  // no () because no side-effect
  println()       // better to not drop the ()
```

> To summarize, it is encouraged style in Scala to define methods that take no parameters and have no side effects as parameterless methods, i.e., leaving off the empty parentheses. On the other hand, you should never define a method that has side-effects without parentheses, because then invocations of that method would look like a field selection. So your clients might be surprised to see the side effects. Similarly, whenever you invoke a function that has side effects, be sure to include the empty parentheses when you write the invocation. Another way to think about this is if the function you're calling performs an operation, use the parentheses, but if it merely provides access to a property, leave the parentheses off.

Reply:
True that it is good practise to use `()` when there is a side effect and omit them when there is no side effect.

But.....

Things get more complicated when the side effect is suspended in `F`.  Here, the `F[Unit]` has no side effects, until the "end of the world"™ when we call `program.unsafeRunSync()` -- and notice the very deliberate parentheses on that function!

I do not have a strong opinion on this, and I have seen it done both ways for suspended effects.  I don't know if there is an agreed best practise for this situation, and I have never seen it discussed anywhere.

Reply:
What you say makes sense, and I also have no strong feelings about it. I ended up keeping the `()` because if the IDE is suggesting it, it will be an ongoing annoyance if left unchanged.

Reply:
I think it might be some IDE bug if it doesn't recognize `F` - I'm in relatively strong opposition to `()`

Reply:
I reverted those changes.

This ended up being the mostly hotly discussed change in the PR; but would you just look at the lovely bike shed we have now. :)